### PR TITLE
Additional index and string default function implementations

### PR DESCRIPTION
**Pull request by aaronbloom**
New functions: `trim`, `length`, `contains`, `startsWith`, `endsWith`, `indexOf`, `lastIndexOf`, `reverse`, `get`.

Add more overloads for `today`.

Made more implementations common between the CSharp and Expression implementations.

Support array formatted output in Repl.

image

### CODE REVIEW

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 1]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("substring('test_substring', 5)", "substring")]
+    [TestCase("substring('test_substring', -5)", null)]
+    [TestCase("substring('test_substring', 0, 4)", "test")]
+    [TestCase("substring('test_substring', 0, 0)", "")]
+    [TestCase("substring('test_substring', 0, 100)", null)]
+    [TestCase("substring('test_substring', 100, 100)", null)]
+    [TestCase("substring('test_substring', -1, 100)", null)]
+    [TestCase("substring('test_substring', -1, -1)", null)]
+    [TestCase("split('test_split', '_')", new[] { "test", "split" })]
+    [TestCase("split('test_split', null)", new string[0])]
```
Comment (bgever) [inline src/Tests/VCEL.Test/FunctionTests.cs:205]: Should this return the original string?

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 2]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("split(null, '_')", new string[0])]
```
Comment (bgever) [inline src/Tests/VCEL.Test/FunctionTests.cs:206]: I would expect these two cases to return null

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 3]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("uppercase('abcd')", "ABCD")]
+    [TestCase("uppercase(null)", null)]
+    [TestCase("substring('test_substring', 5)", "substring")]
+    [TestCase("substring('test_substring', -5)", null)]
+    [TestCase("substring('test_substring', 0, 4)", "test")]
+    [TestCase("substring('test_substring', 0, 0)", "")]
+    [TestCase("substring('test_substring', 0, 100)", null)]
+    [TestCase("substring('test_substring', 100, 100)", null)]
+    [TestCase("substring('test_substring', -1, 100)", null)]
+    [TestCase("substring('test_substring', -1, -1)", null)]
```
Comment (bgever) [inline src/Tests/VCEL.Test/FunctionTests.cs:203]: What if a null input is given to substring?

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 4]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("contains('test_contains', 'test1')", false)]
+    [TestCase("contains(null, 'test1')", false)]
+    [TestCase("startsWith('test_startswith', 'test')", true)]
+    [TestCase("startsWith('test_startswith', 'test1')", false)]
+    [TestCase("startsWith(null, 'test1')", false)]
+    [TestCase("endsWith('test_endswith', 'endswith')", true)]
+    [TestCase("endsWith('test_endswith', 'endswith1')", false)]
+    [TestCase("endsWith('test_endswith', 'test_endswith')", true)]
+    [TestCase("endsWith('test_endswith', 'test')", false)]
+    [TestCase("endsWith(null, 'endswith1')", false)]
```
Comment (bgever) [inline src/Tests/VCEL.Test/FunctionTests.cs:230]: What if the needle is `null`?

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 5]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("replace('test_replace', '_replace', '_test')", "test_test")]
+    [TestCase("replace(null, '_replace', '_test')", null)]
+    [TestCase("replace('test_replace', null, '_test')", "test_replace")]
+    [TestCase("replace('test_replace', '_replace', null)", "test")]
+    [TestCase("replace(null, null, null)", null)]
+    [TestCase("get('test_get', 0)", "t")]
+    [TestCase("get('test_get', 1)", "e")]
+    [TestCase("get('test_get', 6)", "e")]
+    [TestCase("get('test_get', -1)", "t")]
+    [TestCase("get('test_get', -2)", "e")]
```
Comment (bgever) [inline src/Tests/VCEL.Test/FunctionTests.cs:251]: These negative end lookups would be nice for the first positional arg of `substring` as well

File: src/VCEL.CSharp/Expression/Func/DefaultCSharpFunctions.cs [hunk 6]
```diff
@@ -3,116 +3,97 @@
 using System.Linq;
 using VCEL.Core.Expression.Func;
 
-namespace VCEL.CSharp.Expression.Func
+namespace VCEL.CSharp.Expression.Func;
+
+public class DefaultCSharpFunctions : IFunctions<string>
```
Comment (WillLiu-Eclipse) [inline src/VCEL.CSharp/Expression/Func/DefaultCSharpFunctions.cs:8]: looks like we did not added these new function support to javascript vcel?

File: src/Tests/VCEL.Test/FunctionTests.cs [hunk 7]
```diff
@@ -1,384 +1,515 @@
...
+    [TestCase("length([1, 2, 3])", 3)]
+    [TestCase("length([null])", 1)]
+    [TestCase("length(null)", null)]
+    [TestCase("length([...[1], 2)", 2)]
+    [TestCase("contains([1, 2, 3], 1)", true)]
+    [TestCase("contains([1, 2, 3], 4)", false)]
+    [TestCase("contains([1, 2, 3], null)", false)]
+    [TestCase("contains(null, 1)", false)]
+    [TestCase("contains(null, null)", false)]
+    [TestCase("startsWith([1, 2, 3], 1)", true)]
```
Comment (WillLiu-Eclipse) [inline src/Tests/VCEL.Test/FunctionTests.cs:285]: feels a bit weird to use startswith on array.

this together with "split" can do a lot of things

File: src/VCEL.CSharp/Expression/Func/DefaultCSharpFunctions.cs (hunk 6 above)
Comment (aaronbloom) [reply src/VCEL.CSharp/Expression/Func/DefaultCSharpFunctions.cs:8]: Updated with JS function implementations in latest commit.

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 5 above)
Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs:253]: I'll leave as is for now, but agree would be a good change to make

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 4 above)
Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs:231]: Updated

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 3 above)
Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs:203]: Updated

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 2 above)
Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs]: Updated

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 1 above)
Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs]: Updated

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 5 above)
Comment (bgever) [reply src/Tests/VCEL.Test/FunctionTests.cs:253]: Agree, nice to have

**Review by bgever — APPROVED** [review]

File: src/Tests/VCEL.Test/FunctionTests.cs (hunk 7 above)
Comment (bgever) [reply src/Tests/VCEL.Test/FunctionTests.cs:287]: I'm undecided, but wonder what happens when you do `startsWith([1, 2, 3], [1, 2])`.

Comment (aaronbloom) [reply src/Tests/VCEL.Test/FunctionTests.cs:287]: Not supported right now, but I think a possible future improvement

**Pull request merged**